### PR TITLE
Adding IBC wasm hooks

### PR DESCRIPTION
**Pull request by nicolaslara**
This PR adds Osmosis' [IBC wasm hooks](https://github.com/osmosis-labs/osmosis/blob/main/x/ibc-hooks/README.md) to Juno. 

This allows users to execute a contract after an IBC transfer is received by specifying it in the memo field.

### CODE REVIEW

**Comment by faddat** [comment]
Sir, thank you!

**Comment by nicolaslara** [comment]
> I think that requiring v0.46.x when in reality using only 45-series stuff is worth addressing in osmosis.

yeah, that's fixed in the latest branches (main and v14.x). Using the 14.0-rc commit should probably fix it. I can try it out tomorrow morning

**Comment by faddat** [comment]
oh, if it's that simple, I'll have a poke at it now.  Thank you!

**Review by faddat — APPROVED** [review]
Approved for uni to bump before mainnet

**Comment by faddat** [comment]
Hey sorry this "worked on my machine"

**Comment by nicolaslara** [comment]
changed the dependency to the latest release. 

The problem here is that osmoutils is using https://github.com/osmosis-labs/cosmos-sdk/blob/osmosis-main/client/context.go#L352 from the osmosis sdk fork, which now has a different signature from the old one.

I think the easiest solution is to remove the dependency on osmoutils for ibc-hooks. I'll push a branch with it soon

**Comment by nicolaslara** [comment]
This should be fixed now. Tests pass locally.

**Comment by faddat** [comment]
Well thank you very much

**Comment by nicolaslara** [comment]
I'm running into a new issue after main being merged into this, that I think might be because of wasm 0.30.

```
runtime/debug.Stack()
	runtime/debug/stack.go:24 +0x64
github.com/cosmos/cosmos-sdk/baseapp.newDefaultRecoveryMiddleware.func1({0x1e2e020, 0x4003c7e630})
	github.com/cosmos/cosmos-sdk@v0.45.11/baseapp/recovery.go:71 +0x24
github.com/cosmos/cosmos-sdk/baseapp.newRecoveryMiddleware.func1({0x1e2e020?, 0x4003c7e630?})
	github.com/cosmos/cosmos-sdk@v0.45.11/baseapp/recovery.go:39 +0x34
github.com/cosmos/cosmos-sdk/baseapp.processRecovery({0x1e2e020, 0x4003c7e630}, 0x400168a0c0?)
	github.com/cosmos/cosmos-sdk@v0.45.11/baseapp/recovery.go:28 +0x38
github.com/cosmos/cosmos-sdk/baseapp.processRecovery({0x1e2e020, 0x4003c7e630}, 0x2c83240?)
	github.com/cosmos/cosmos-sdk@v0.45.11/baseapp/recovery.go:33 +0x60
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runTx.func1()
	github.com/cosmos/cosmos-sdk@v0.45.11/baseapp/baseapp.go:632 +0xb0
panic({0x1e2e020, 0x4003c7e630})
	runtime/panic.go:884 +0x20c
github.com/cosmos/cosmos-sdk/store/cachemulti.Store.GetKVStore({{0x2c81430, 0x4003b1dc00}, 0x40038b7f80, 0x0, {0x0, 0x0}, 0x0, 0x40038b7e60}, {0x0, 0x0})
	github.com/cosmos/cosmos-sdk@v0.45.11/store/cachemulti/store.go:188 +0xd8
github.com/cosmos/cosmos-sdk/types.Context.KVStore({{0x2c71c68, 0x4000056350}, {0x2c83240, 0x4000c224c0}, {{0xb, 0x0}, {0x4005084a70, 0x9}, 0xb5, {0x10f323d6, ...}, ...}, ...}, ...)
	github.com/cosmos/cosmos-sdk@v0.45.11/types/context.go:253 +0x68
github.com/CosmWasm/wasmd/x/wasm/keeper.Keeper.contractInstance({{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, ...}, ...)
	github.com/CosmWasm/wasmd@v0.30.0/x/wasm/keeper/keeper.go:759 +0xac
github.com/CosmWasm/wasmd/x/wasm/keeper.Keeper.execute({{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, ...}, ...)
	github.com/CosmWasm/wasmd@v0.30.0/x/wasm/keeper/keeper.go:404 +0x158
github.com/CosmWasm/wasmd/x/wasm/keeper.PermissionedKeeper.Execute(...)
	github.com/CosmWasm/wasmd@v0.30.0/x/wasm/keeper/contract_keeper.go:95
github.com/CosmWasm/wasmd/x/wasm/keeper.msgServer.ExecuteContract({{0x2c82138?, 0x4001186300?}}, {0x2c71cd8, 0x4003c4ff20}, 0x4003024e88)
	github.com/CosmWasm/wasmd@v0.30.0/x/wasm/keeper/msg_server.go:139 +0x5f4
github.com/osmosis-labs/osmosis/x/ibc-hooks.WasmHooks.execWasmMsg({_, _, {_, _}}, {{0x2c71c68, 0x4000056350}, {0x2c83240, 0x4000c224c0}, {{0xb, 0x0}, ...}, ...}, ...)
	github.com/osmosis-labs/osmosis/x/ibc-hooks@v0.0.0-20230110104305-322e8478dbe8/wasm_hook.go:132 +0x1bc
github.com/osmosis-labs/osmosis/x/ibc-hooks.WasmHooks.OnRecvPacketOverride({_, _, {_, _}}, {{_, _}, _}, {{0x2c71c68, 0x4000056350}, {0x2c83240, ...}, ...}, ...)
	github.com/osmosis-labs/osmosis/x/ibc-hooks@v0.0.0-20230110104305-322e8478dbe8/wasm_hook.go:113 +0x464
github.com/osmosis-labs/osmosis/x/ibc-hooks.IBCMiddleware.OnRecvPacket({{_, _}, _}, {{0x2c71c68, 0x4000056350}, {0x2c83240, 0x4000c224c0}, {{0xb, 0x0}, {0x4005084a70, ...}, ...}, ...}, ...)
	github.com/osmosis-labs/osmosis/x/ibc-hooks@v0.0.0-20230110104305-322e8478dbe8/ibc_module.go:176 +0x300
github.com/cosmos/ibc-go/v4/modules/core/keeper.Keeper.RecvPacket({{0x0, 0x0}, {0x2c813c0, 0x4000c99bf0}, {{0x2c5bbc0, 0x4001236d70}, {0x2c813c0, 0x4000c99bf0}, {{0x2c813c0, 0x4000c99bf0}, ...}, ...}, ...}, ...)
	github.com/cosmos/ibc-go/v4@v4.2.0/modules/core/keeper/msg_server.go:413 +0x6e4
github.com/cosmos/ibc-go/v4/modules/core/04-channel/types._Msg_RecvPacket_Handler.func1({0x2c71cd8, 0x4003abbfb0}, {0x209c020?, 0x4003284160})
	github.com/cosmos/ibc-go/v4@v4.2.0/modules/core/04-channel/types/tx.pb.go:1322 +0x74
github.com/cosmos/cosmos-sdk/baseapp.(*MsgServiceRouter).RegisterService.func2.1({0x2c71cd8, 0x4003abbf80}, {0x4003027938?, 0xda6000?}, 0x211a4a0?, 0x4000e8ad98)
	github.com/cosmos/cosmos-sdk@v0.45.11/baseapp/msg_service_router.go:113 +0xc0
github.com/cosmos/ibc-go/v4/modules/core/04-channel/types._Msg_RecvPacket_Handler({0x210f2e0?, 0x4000ea9680}, {0x2c71cd8, 0x4003abbf80}, 0x29fa820, 0x40041909a0)
	github.com/cosmos/ibc-go/v4@v4.2.0/modules/core/04-channel/types/tx.pb.go:1324 +0x134
github.com/cosmos/cosmos-sdk/baseapp.(*MsgServiceRouter).RegisterService.func2({{0x2c71c68, 0x4000056350}, {0x2c83240, 0x4003b1c640}, {{0xb, 0x0}, {0x4005084a70, 0x9}, 0xb5, {0x10f323d6, ...}, ...}, ...}, ...)
	github.com/cosmos/cosmos-sdk@v0.45.11/baseapp/msg_service_router.go:126 +0x31c
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runMsgs(_, {{0x2c71c68, 0x4000056350}, {0x2c83240, 0x4003b1c640}, {{0xb, 0x0}, {0x4005084a70, 0x9}, 0xb5, ...}, ...}, ...)
	github.com/cosmos/cosmos-sdk@v0.45.11/baseapp/baseapp.go:760 +0x464
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runTx(0x4000d60000, 0x2, {0x400564f500, 0x91f, 0xa80})
	github.com/cosmos/cosmos-sdk@v0.45.11/baseapp/baseapp.go:717 +0x8e0
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).Simulate(0x10?, {0x400564f500?, 0x1?, 0x0?})
	github.com/cosmos/cosmos-sdk@v0.45.11/baseapp/test_helpers.go:23 +0x30
github.com/cosmos/cosmos-sdk/x/auth/tx.txServer.Simulate({{{0x0, 0x0, 0x0}, {0x2c8fd48, 0x4000fc6d00}, {0x0, 0x0}, {0x2c78590, 0x4000c99bf0}, {0x2c869b8, ...}, ...}, ...}, ...)
	github.com/cosmos/cosmos-sdk@v0.45.11/x/auth/tx/service.go:120 +0x118
github.com/cosmos/cosmos-sdk/types/tx._Service_Simulate_Handler.func1({0x2c71cd8, 0x4000bbc300}, {0x207ecc0?, 0x40039dbc00})
	github.com/cosmos/cosmos-sdk@v0.45.11/types/tx/service.pb.go:917 +0x74
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).RegisterGRPCServer.func1({0x2c71cd8, 0x4005072960}, {0x207ecc0, 0x40039dbc00}, 0x400102b998?, 0x40042fb3c8)
	github.com/cosmos/cosmos-sdk@v0.45.11/baseapp/grpcserver.go:66 +0x320
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1({0x2c71cd8?, 0x4005072960?}, {0x207ecc0?, 0x40039dbc00?})
	github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x3c
github.com/grpc-ecosystem/go-grpc-middleware/recovery.UnaryServerInterceptor.func1({0x2c71cd8?, 0x4005072960?}, {0x207ecc0?, 0x40039dbc00?}, 0xba18?, 0x40039dbc60?)
	github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/recovery/interceptors.go:33 +0x98
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1({0x2c71cd8?, 0x4005072960?}, {0x207ecc0?, 0x40039dbc00?})
	github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:25 +0x3c
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1({0x2c71cd8, 0x4005072960}, {0x207ecc0, 0x40039dbc00}, 0x400102ba78?, 0xda5448?)
	github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:34 +0xb8
github.com/cosmos/cosmos-sdk/types/tx._Service_Simulate_Handler({0x2030f60?, 0x4000c27a00}, {0x2c71cd8, 0x4005072960}, 0x400460e900, 0x4005072990)
	github.com/cosmos/cosmos-sdk@v0.45.11/types/tx/service.pb.go:919 +0x134
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).RegisterGRPCServer.func2({0x2030f60, 0x4000c27a00}, {0x2c71cd8, 0x4005072960}, 0x20dc4e0?, 0x4005690e00?)
	github.com/cosmos/cosmos-sdk@v0.45.11/baseapp/grpcserver.go:80 +0xd0
google.golang.org/grpc.(*Server).processUnaryRPC(0x4000fa5340, {0x2c7e980, 0x4002e50300}, 0x4005690e00, 0x4000db5680, 0x400049d400, 0x0)
	google.golang.org/grpc@v1.50.1/server.go:1210 +0xac0
google.golang.org/grpc.(*Server).handleStream(0x4000fa5340, {0x2c7e980, 0x4002e50300}, 0x4005690e00, 0x0)
	google.golang.org/grpc@v1.50.1/server.go:1533 +0x840
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	google.golang.org/grpc@v1.50.1/server.go:871 +0x84
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.50.1/server.go:869 +0x294
: panic [cosmos/cosmos-sdk@v0.45.11/baseapp/recovery.go:69] With gas wanted: '0' and gas used: '195226' ", details: [], metadata: MetadataMap { headers: {"content-type": "application/grpc", "x-cosmos-block-height": "181"} }
```

For reference. I'm using osmosis/localrelayer from [this branch](https://github.com/osmosis-labs/osmosis/tree/nicolas/test-for-juno/tests/localrelayer) and a [script](https://github.com/osmosis-labs/osmosis/blob/nicolas/test-for-juno/tests/ibc-hooks/test_hooks.sh) to do manual testing of this.

I'll review this and report back, but if anyone has pointers, just let me know

**Comment by nicolaslara** [comment]
> Osmosis 14 is supposed to use WASM 29? 

we're still on wasm29, yes. 

Might need to do some compatibility work there. I think juno was on wasm 29 when I first wrote this PR

**Comment by nicolaslara** [comment]
Ignore my comment above. This was not at all related to the wasm version (they are interface compatible). 

The issue was that I was initializing the hooks in the wrong order and thus the contract keeper was empty in it. This has been fixed in my latest commit.

File: app/keepers/keepers.go [hunk 1]
```diff
@@ -421,6 +445,10 @@ func NewAppKeepers(
 	appKeepers.ScopedICAControllerKeeper = scopedICAControllerKeeper
 	appKeepers.ScopedInterTxKeeper = scopedInterTxKeeper
 
+	// set the contract keeper for the Ics20WasmHooks
+	appKeepers.ContractKeeper = wasmkeeper.NewDefaultPermissionKeeper(appKeepers.WasmKeeper)
```
Comment (JakeHartnell) [inline app/keepers/keepers.go:449]: So these hooks are governed by permissioned contracts? Trying to understand how this works in a permissionless smart contract context.

**Comment by Reecepbcups** [comment]
e2e so flakey if multiple CI are running. probably need to randomize ports on start

Will review this tomorrow

File: app/keepers/keepers.go (hunk 1 above)
Comment (nicolaslara) [reply app/keepers/keepers.go:449]: `NewDefaultPermissionKeeper` returns a wrapper around the wasm keeper with the "default" permissions. But the permissions here only matter for create, instantiate(2), migrate, UpdateContractAdmin, ClearContractAdmin, and SetAccessConfig. None of which we're using.

Even if we were using those, I think the default permissions is what we want. It's used in places like [this](https://github.com/CosmWasm/wasmd/blob/558147d054c88840a804a3ec4e0bba5b3cb5905a/x/wasm/keeper/keeper.go#L190-L203) and define like [this](https://github.com/CosmWasm/wasmd/blob/558147d054c88840a804a3ec4e0bba5b3cb5905a/x/wasm/keeper/authz_policy.go#L27-L45). Plus, the only other policy is "gov" which just allows everything (and we still want, for example, that only admins can migrate contracts).

An alternative here would be to write our own wrapper that implements [ContractOpsKeeper](https://github.com/CosmWasm/wasmd/blob/558147d054c88840a804a3ec4e0bba5b3cb5905a/x/wasm/types/exported_keepers.go#L28) that panics on the functions we're not using, but I don't see the point when wasmd already implements this in [PermissionedKeeper](https://github.com/CosmWasm/wasmd/blob/558147d054c88840a804a3ec4e0bba5b3cb5905a/x/wasm/keeper/contract_keeper.go#L37).

**Comment by faddat** [comment]
Yeah you're correct about that we have a test flaking out for sure. I'm not sure that that's the solution but if it is that's super.

**Review by Reecepbcups — APPROVED** [review]
I have tested it locally using an updated test_node & relayer scripts found here in PR #492, worked perfectly.

Will add better test in the future via e2e & fix up those things after v12 on testnet

 :) ty

File: app/modules.go [hunk 2]
```diff
@@ -140,8 +143,9 @@ func appModules(
 		feeshare.NewAppModule(app.FeeShareKeeper, app.AccountKeeper),
 		wasm.NewAppModule(appCodec, &app.WasmKeeper, app.StakingKeeper, app.AccountKeeper, app.BankKeeper),
 		ica.NewAppModule(&app.ICAControllerKeeper, &app.ICAHostKeeper),
-		intertx.NewAppModule(appCodec, app.InterTxKeeper),
+		intertx.NewAppModule(appCodec, app.InterTxKeeper),
```
Comment (Reecepbcups) [inline app/modules.go:146]:
```suggestion
		intertx.NewAppModule(appCodec, app.InterTxKeeper),
```

**Review by faddat — APPROVED** [review]

**Review by Reecepbcups — APPROVED** [review]

**Pull request merged**